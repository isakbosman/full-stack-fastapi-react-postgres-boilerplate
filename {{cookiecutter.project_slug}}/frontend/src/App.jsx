import { hot } from 'react-hot-loader/root';
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Router, Switch, Route } from 'react-router-dom';
import { Helmet } from 'react-helmet-async';
import styled, { css, ThemeProvider } from 'styled-components';
import treeChanges from 'tree-changes';

import history from 'modules/history';
import theme, { headerHeight } from 'modules/theme';
import { utils } from 'styled-minimal';

import config from 'config';
import { showAlert } from 'actions';

import Home from 'routes/Home';
import Private from 'routes/Private';
import NotFound from 'routes/NotFound';

import Header from 'components/Header';
import SystemAlerts from 'components/SystemAlerts';

import Footer from 'components/Footer';
import GlobalStyles from 'components/GlobalStyles';
import RoutePublic from 'components/RoutePublic';
import RoutePrivate from 'components/RoutePrivate';

const AppWrapper = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  opacity: 1 !important;
  position: relative;
  transition: opacity 0.5s;
`;

const MainPrivate = ({ isAuthenticated }) =>
  isAuthenticated &&
  css`
    padding: ${utils.px(headerHeight)} 0 0;
  `;

const Main = styled.main`
  min-height: 100vh;

  ${MainPrivate};
`;

export class App extends React.Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
  };

  componentDidUpdate(prevProps) {
    const { dispatch } = this.props;
    const { changedTo } = treeChanges(prevProps, this.props);

    /* istanbul ignore else */
    if (changedTo('user.isAuthenticated', true)) {
      dispatch(showAlert('Hello! And welcome!', { variant: 'success', icon: 'bell' }));
    }
  }

  render() {
    const { dispatch, user } = this.props;

    return (
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <AppWrapper logged={user.isAuthenticated}>
            <Helmet
              defer={false}
              htmlAttributes={{ lang: 'pt-br' }}
              encodeSpecialCharacters={true}
              defaultTitle={config.name}
              titleTemplate={`%s | ${config.name}`}
              titleAttributes={{ itemprop: 'name', lang: 'pt-br' }}
            />
            {user.isAuthenticated && <Header dispatch={dispatch} user={user} />}
            <Main isAuthenticated={user.isAuthenticated}>
              <Switch>
                <RoutePublic
                  isAuthenticated={user.isAuthenticated}
                  path="/"
                  exact
                  component={Home}
                />
                <RoutePrivate
                  isAuthenticated={user.isAuthenticated}
                  path="/private"
                  component={Private}
                />
                <Route component={NotFound} />
              </Switch>
            </Main>
            <Footer />
            <SystemAlerts />
            <GlobalStyles />
          </AppWrapper>
        </ThemeProvider>
      </Router>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return {
    user: state.user,
  };
}

export default hot(connect(mapStateToProps)(App));
