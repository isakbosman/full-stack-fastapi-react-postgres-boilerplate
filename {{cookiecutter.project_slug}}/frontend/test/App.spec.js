import React from 'react';

import { App } from 'App';

const mockDispatch = jest.fn();

const props = {
  app: {
    alerts: [],
  },
  dispatch: mockDispatch,
  user: {
    isAuthenticated: false,
  },
};

function setup(ownProps = props) {
  return shallow(<App {...ownProps} />, { attachTo: document.getElementById('react') });
}

describe('App', () => {
  const wrapper = setup();

  it('should render properly for anonymous users', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should render properly for logged users', () => {
    wrapper.setProps({
      ...wrapper.props(),
      user: {
        isAuthenticated: true,
      },
    });

    expect(wrapper.find('Header')).toExist();
  });
});
