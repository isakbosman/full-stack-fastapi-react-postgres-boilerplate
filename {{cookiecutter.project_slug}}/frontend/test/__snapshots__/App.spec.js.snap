// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render properly for anonymous users 1`] = `
<Router history="{{...}}">
  <ThemeProvider theme="{{...}}">
    <AppWrapper logged="{false}">
      <Helmet defer="{false}"
              htmlattributes="{{...}}"
              encodespecialcharacters="{true}"
              defaulttitle="React Redux Saga Boilerplate"
              titletemplate="%s | React Redux Saga Boilerplate"
              titleattributes="{{...}}"
      >
      </Helmet>
      <Main isauthenticated="{false}">
        <Switch>
          <RoutePublic isauthenticated="{false}"
                       path="/"
                       exact="{true}"
                       component="{{...}}"
                       to="/private"
          >
          </RoutePublic>
          <RoutePrivate isauthenticated="{false}"
                        path="/private"
                        component="{[Function:"
                        private]}
                        to="/"
          >
          </RoutePrivate>
          <Route component="{[Function:"
                 notfound]}
          >
          </Route>
        </Switch>
      </Main>
      <Footer>
      </Footer>
      <Connect(SystemAlerts)>
      </Connect(SystemAlerts)>
      <_default>
      </_default>
    </AppWrapper>
  </ThemeProvider>
</Router>
`;
